libtcmu_aio.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright 2017, Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#define _GNU_SOURCE
+#include <errno.h>
+#include <assert.h>
+#include <stdint.h>
+#include <pthread.h>
+
+#include "libtcmu.h"
+#include "libtcmu_priv.h"
+#include "libtcmu_aio.h"
+#include "tcmu-runner.h"
+
+void tcmulib_track_aio_request_start(struct tcmu_device *dev)
+{
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	pthread_cleanup_push(_cleanup_spin_lock, (void *)&aio_track->track_lock);
+	pthread_spin_lock(&aio_track->track_lock);
+
+	++aio_track->tracked_aio_ops;
+
+	pthread_spin_unlock(&aio_track->track_lock);
+	pthread_cleanup_pop(0);
+}
+
+void tcmulib_track_aio_request_finish(struct tcmu_device *dev, int *is_idle)
+{
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	pthread_cleanup_push(_cleanup_spin_lock, (void *)&aio_track->track_lock);
+	pthread_spin_lock(&aio_track->track_lock);
+
+	assert(aio_track->tracked_aio_ops > 0);
+
+	--aio_track->tracked_aio_ops;
+	if (is_idle) {
+		*is_idle = (aio_track->tracked_aio_ops == 0) ? 1 : 0;
+	}
+
+	pthread_spin_unlock(&aio_track->track_lock);
+	pthread_cleanup_pop(0);
+}
+
+static void tcmulib_callout_finished(struct tcmu_device *dev,
+				     struct tcmulib_cmd *cmd, int ret)
+{
+	cmd->callout_cbk(dev, cmd, ret);
+}
+
+static int call_stub_exec_async(struct tcmu_device *dev,
+				struct tcmulib_cmd *cmd,
+				struct tcmu_call_stub *stub)
+{
+	int rc;
+	switch(stub->sop) {
+	case TCMU_STORE_OP_READ:
+	case TCMU_STORE_OP_WRITE:
+		rc = (int) stub->u.rw.exec(dev, stub->u.rw.iov,
+					   stub->u.rw.iov_cnt, stub->u.rw.off);
+		break;
+	case TCMU_STORE_OP_FLUSH:
+		rc = stub->u.flush.exec(dev);
+		break;
+	case TCMU_STORE_OP_HANDLE_CMD:
+		rc = stub->u.handle_cmd.exec(dev, cmd);
+		break;
+	default:
+		tcmu_err("unhandled store operation\n");
+		assert(0 == "unhandled store operation");
+	}
+
+	return rc;
+}
+
+static void call_stub_exec_sync(struct tcmu_device *dev,
+				struct tcmulib_cmd *cmd,
+				struct tcmu_call_stub *stub)
+{
+	ssize_t ret;
```

Comment:
Do we need both of these because you are having the store callouts return different values? I do not think we need to. Bryant and I have converted qcow and file* to use the store callouts and to work similarly.

You could have them all return normal old ints. -EXYZ == error. 0 means it was successfully completed or queued for the async case. It seems we fail if the entire IO cannot be completed so we do not need to return the number of bytes read/written, but we could return the number of bytes in the sync case if we needed it.

Or.

You can have them all return SCSI and TCMU values. The handler calls tcmu_set_sense_data on failure.

It does not matter to me. I like handlers not having to deal with SCSI details, but the second option is more flexibly. Any preference?